### PR TITLE
Improve Name.lastIndexOf

### PR DESCRIPTION
**Pull request by som-snytt**
Perform the test instead of creating a string to
delegate to. We have access to the underlying Name
array; if testing for short strings, the penalty
for indexing into them is small.

Maybe improves https://github.com/scala/scala/pull/8928 which avoids incurring the allocation by probing for a character.

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Names.scala [hunk 1]
```diff
@@ -398,11 +398,28 @@ trait Names extends api.Names {
...
+      if (s.isEmpty()) return length
+
+      val last = s(s.length - 1)
+      var i = length
+      while (s.length() <= i) {
+        if (_chrs(start + i - 1) == last) {
+          var j = 2
+          while (j <= s.length() && _chrs(start + i - j) == s.charAt(s.length() - j))
+            j += 1
+          if (j > s.length()) return i - s.length()
```
Comment (dwijnand) [inline src/reflect/scala/reflect/internal/Names.scala]: Compute `s.length()` once?

File: src/reflect/scala/reflect/internal/Names.scala [hunk 2]
```diff
@@ -398,11 +398,28 @@ trait Names extends api.Names {
       false
     }
 
+    def lastIndexOf(s: String): Int = {
```
Comment (mkeskells) [inline src/reflect/scala/reflect/internal/Names.scala]: how does this compare to java implemention in `String`/`StringBuilder` - 
```
static int lastIndexOf(char[] source, int sourceOffset, int sourceCount,
            char[] target, int targetOffset, int targetCount,
            int fromIndex) {
```

Comment (som-snytt) [reply src/reflect/scala/reflect/internal/Names.scala]: If someone chimes in that this is a good idea, I'll benchmark and optimise. This is a first cut at something reasonable. We're comparing a String (which may be short um `"$$"`) with our array.

File: src/reflect/scala/reflect/internal/Names.scala (hunk 1 above)
Comment (som-snytt) [reply src/reflect/scala/reflect/internal/Names.scala]: The real problem is that use of parens is inconsistent.

**Comment by som-snytt** [comment]
No one said this was a good idea.

File: src/reflect/scala/reflect/internal/Names.scala (hunk 2 above)
Comment (som-snytt) [reply src/reflect/scala/reflect/internal/Names.scala]: I guess there are only so many ways to look for a string backwards.

**Review by dwijnand — APPROVED** [review]

**Pull request merged**